introduce PageObjectRegistry with @hande_urls annotations

BurnzZ opened the pull request:
This is built as a light-weight version of the new features introduce by this PR: https://github.com/scrapinghub/web-poet/pull/16/files

In contrast, this only contains a minimal set of features that allows developers to:
- import and use the `default_registry`'s `@handle_urls` annotation for declaring `OverrideRule`s.
- use `search_overrides()` to find specific `OverrideRule`s amased from importing multiple Page Object Projects (POP).

This means that other non-essential features from the old PR aren't included:
- `registry_pool` which allows keeping track of other non-default `PageObjectRegistry` instances
- CLI-tool for displaying all of the `OverrideRule`s
- `copy_overrides_from()` which allows copying of `OverrideRule`s from multiple non-default `PageObjectRegistry` instances
- `remove_overrides()` since an **exclusion-list** strategy may be brittle if the underlying source changes. This results in some unintentional `OverrideRules` to be present in the output if the developer isn't aware of them.
- `replace_overrides()`  since a more explicit alternative would be creating a new `OverrideRule` which has the intended attribute replacements.
- the `filter` param of `get_overrides()` since it could break the developer's expectation if new `OverrideRule`s are added to the filter module which could lead to unintentional `OverrideRules` to be present.

### Progress:

* [x] Tests
* [x] rewrite docs to focus on more minimal usage

### To address in other PRs:

* POP entry point - https://github.com/scrapinghub/web-poet/pull/28
* `OverrideRule.instead_of` to support data types as well

kmike commented on docs/intro/overrides.rst:
```diff
@@ -0,0 +1,349 @@
+----------------------------------
+
+The :meth:`~.PageObjectRegistry.get_overrides` method from the ``web_poet.default_registry``
+allows discovery and retrieval of  all :class:`~.OverrideRule` from your project.
```
I think get_overrides is not doing discovery, and also it doesn't limit rules to a certain project:

```suggestion
allows retrieval of all :class:`~.OverrideRule` in the registry.
```

BurnzZ replied:
Good point. Updated this on https://github.com/scrapinghub/web-poet/pull/27/commits/37e5ba1a3696f629801478fe78ea1729164f39df

kmike commented on docs/intro/overrides.rst:
```diff
@@ -0,0 +1,349 @@
+
+    :meth:`~.PageObjectRegistry.get_overrides` relies on the fact that all essential
+    packages/modules which contains the :func:`web_poet.handle_urls`
+    annotations are properly loaded.
```
```suggestion
    annotations are properly loaded (i.e. imported).
```

kmike replied:
Overall, I think it may be a bit more clear if in some places further in the docs instead of "loaded" is replaced with "imported". consume_modules is not doing anything fancy, it just imports modules.

BurnzZ replied:
Updated this on https://github.com/scrapinghub/web-poet/pull/27/commits/37e5ba1a3696f629801478fe78ea1729164f39df

kmike commented on docs/intro/overrides.rst:
```diff
@@ -0,0 +1,349 @@
+        rules = default_registry.get_overrides()
+
+        # Fortunately, `get_overrides()` provides a shortcut for the lines above:
+        rules = default_registry.get_overrides(consume=["external_package_A.po", "another_ext_package.lib"])
```
Because of the caveats I'm not sure it makes sense to provide `consume` keyword for get_overrides - the code above reads as it'd return only POs from the listed modules. With 

```py
consume_modules("external_package_A.po", "another_ext_package.lib")
rules = default_registry.get_overrides()
```

the caveat is more clear IMHO, and it's the same amount of code. Do you think we can remove this method?

BurnzZ replied:
Yeah, we can make it more explicit by having to access `consume_modules()` externally.

Removed this on https://github.com/scrapinghub/web-poet/pull/27/commits/37e5ba1a3696f629801478fe78ea1729164f39df

kmike commented on docs/intro/overrides.rst:
```diff
@@ -0,0 +1,349 @@
+Suppose that the use case from the previous section has changed wherein a
+subset of :class:`~.OverrideRule` would be used. This could be achieved by
+using the :meth:`~.PageObjectRegistry.search_overrides` method which allows for
+convenient selection of a subset of rules from the ``default_registry``.
```
```suggestion
convenient selection of a subset of rules from the registry.
```

BurnzZ replied:
Updated this on https://github.com/scrapinghub/web-poet/pull/27/commits/37e5ba1a3696f629801478fe78ea1729164f39df

kmike commented on docs/intro/overrides.rst:
```diff
@@ -0,0 +1,349 @@
+
+There are two main ways we recommend in solving this.
+
+**1. Priority Resolution**
```
I think the main use case here could be the following:

1. There is a library of page objects
2. In your project you want to have a different version for one of POs available in the library

In this case there is no need to modify any override rules, it's a matter of using handle_urls decorator in your own project, and setting a priority higher than the default.

BurnzZ replied:
@kmike to clarify, are you suggesting something like the following?

If we have the conflict like:

```python

rules = default_registry.get_overrides()
print(rules)
# OverrideRule(for_patterns=Patterns(include=['site_2.com'], exclude=[], priority=500), use=<class 'ecommerce_page_objects.site_2.EcomSite2'>, instead_of=<class 'common_items.ProductGenericPage'>, meta={})
# OverrideRule(for_patterns=Patterns(include=['site_2.com'], exclude=[], priority=500), use=<class 'gadget_sites_page_objects.site_2.GadgetSite2'>, instead_of=<class 'common_items.ProductGenericPage'>, meta={})
```

It can be fixed in the project as something like:

```python
import handle_urls

handle_urls(ProductGenericPage, overrides=ProductGenericPage, priority=600)(EcomSite2)
```

At the moment, this won't work since the `PageObjectRegistry` stores the PO class as the key, and if a rule is already existing for a given PO, it ignores new rules attempted to be registered. [Reference](https://github.com/scrapinghub/web-poet/pull/27/files#diff-409cb5fe79e7d7d5ede41109573e31bd847a147a966f477de80c99279bcdf479R153-R161).

Do you think it's worth refactoring this from `Dict[<PO class>, OverrideRule]` into `Dict[<PO class>, List[OverrideRule]]` instead?

Though this would affect the other implementations like `search_overrides()` and how it's being read in **scrapy-poet** as well. Nonetheless, we're still quite early in the process and we can still accommodate this change.

BurnzZ replied:
@kmike that makes sense! Thanks for explaining it as I initially had the wrong idea. This use case is indeed a common one and it deserves it's own section which I've added in https://github.com/scrapinghub/web-poet/pull/27/commits/7944956802e8c38d939c42b7866d2915f853b1fe.

kmike commented on docs/intro/overrides.rst:
```diff
@@ -0,0 +1,349 @@
+were recently added. This could lead to a `silent-error` of receiving a different
+set of rules than expected.
+
+For this approach, you can use the :meth:`~.PageObjectRegistry.search_overrides`
```
Hm, I was thinking about importing POs directly. How would you do inclusion in search_overrides?

BurnzZ replied:
Added another in-depth example for `search_overrides()` in https://github.com/scrapinghub/web-poet/pull/27/commits/74573319c7b495dbf16dc615be846af51721d508.

kmike commented on web_poet/overrides.py:
```diff
@@ -0,0 +1,305 @@
+
+        # Short-circuit operation if "use" is the only search param used, since
+        # we know that it's being used as the dict key.
+        if set(["use"]) == kwargs.keys():
```
```suggestion
        if {"use"} == kwargs.keys():
```

kmike commented on web_poet/overrides.py:
```diff
@@ -0,0 +1,305 @@
+
+        - ``my_page_obj_project`` `(since it's the same module as the file above)`
+        - ``other_external_pkg.po``
+        - ``another_pkg.lib``
```
+ any other modules imported in the same process. I think the example only works if the code above is a whole script, not a module within a larger package.

BurnzZ replied:
Good one. Updated this on https://github.com/scrapinghub/web-poet/pull/27/commits/37e5ba1a3696f629801478fe78ea1729164f39df

kmike commented on web_poet/overrides.py:
```diff
@@ -0,0 +1,305 @@
+
+    If the ``default_registry`` had other ``@handle_urls`` annotations outside
+    of the packages/modules listed above, then the corresponding
+    :class:`~.OverrideRule` won't be returned.
```
Again, this is only true if they're not imported in some other way. 

I think it may be fine to just explain that consume_module imports modules recursively.

BurnzZ replied:
Updated this in https://github.com/scrapinghub/web-poet/pull/27/commits/37e5ba1a3696f629801478fe78ea1729164f39df

kmike commented:
@BurnzZ the PR looks good to me. There are some ways to improve the docs, but it can be done later. 

+1 to merge after removing "consume" option from get_overrides method.

BurnzZ commented on web_poet/overrides.py:
```diff
@@ -0,0 +1,305 @@
+        return hash((self.for_patterns, self.use, self.instead_of))
+
+
+def _as_list(value: Optional[Strings]) -> List[str]:
```
Moved it in https://github.com/scrapinghub/web-poet/pull/27/commits/50c0046daca9dae90ba0ed1223d2ea91e98768df

BurnzZ commented:
Thanks for the review @kmike ! I've addressed all of your comments alongside the specific commit for the change. Let me know if there's anything left to update here. 🙏

kmike commented on web_poet/utils.py:
```diff
@@ -15,3 +16,22 @@ def new_method(self, *args, **kwargs):
+    """
+    if value is None:
+        return []
+    if not isinstance(value, list):
```
It'd be nice to add a test for tuple, e.g. what should `as_list(("foo", "bar", 123))` return? Should it be a single-item list, or a tuple converted to a list?

BurnzZ replied:
Good point! After looking into this further, I've added more input types that `as_list()` should also be able to handle and convert. Updated in https://github.com/scrapinghub/web-poet/pull/27/commits/5eab374d5767fe2fed821351a5c10fd1e9c9631e.

kmike commented:
Thanks @BurnzZ!
